### PR TITLE
Fix Issue #73, Begin Unboxing at key

### PR DESCRIPTION
**Pull request by clayellis**
I've provided a solution to Issue #73, #54  by allowing for an Unboxer to be created starting at a specified key. Includes keyPath support.

### CODE REVIEW

**Comment by clayellis** [comment]
@JohnSundell, what are your thoughts on this? Do you think this fits in the framework the way you would like? I'd love to collaborate on this feature and find the best way to implement it.

**Comment by JohnSundell** [comment]
@clayellis Sorry I haven't looked at this yet, has been a crazily busy week. Will take a look tomorrow and let you know my thoughts :) Thanks for contributing!

**Comment by clayellis** [comment]
@JohnSundell, awesome! No worries—I know how it is. I hope that means good things are going down at Spotify ;) Good hearing from you. Happy to contribute.

File: README.md [hunk 1]
```diff
@@ -233,28 +233,30 @@ You can also use key paths (for both dictionary keys and array indexes) to unbox
...
-        "running": {
-            "distance": 300
-        }
-    },
-    "devices": [
-        "Macbook Pro",
-        "iPhone",
-        "iPad"
-    ]
+	"name": "John",
```
Comment (JohnSundell) [inline README.md]: Why this change? :)

File: README.md [hunk 2]
```diff
@@ -264,6 +266,56 @@ struct User: Unboxable {
...
+You can also use key paths to directly unbox nested JSON structures. This is useful when you only need to extract a specific object (or objects) out of the JSON body.
+
+```json
+{
+    "company": {
+        "name": "Spotify",
+	},
+	"jobOpenings": [
+		{
+			"title": "Swift Developer",
```
Comment (JohnSundell) [inline README.md]: This indentation is a bit off, we can always fix this post merge thou.

File: Sources/Unbox.swift [hunk 3]
```diff
@@ -872,20 +870,82 @@ private extension Unboxer {
...
+    static func unboxer(at key: String, in dictionary: UnboxableDictionary, isKeyPath: Bool, context: Any?) throws -> Unboxer {
+        let values = try retrieveValues(from: dictionary, at: key, isKeyPath: isKeyPath)
+        let dictionary = values.dictionary
+        let array = values.array
+        let modifiedKey = values.modifiedKey
+        
+        if let dictionary = dictionary[modifiedKey] as? UnboxableDictionary {
+            return Unboxer(dictionary: dictionary, context: context)
+        } else if let array = array, let index = Int(modifiedKey) where index < array.count , let dictionary = array[index] as? UnboxableDictionary {
+            print(dictionary)
```
Comment (JohnSundell) [inline Sources/Unbox.swift]: Remove this :)

**Comment by JohnSundell** [comment]
@clayellis Indeed good things are going down :) Finally, I've had time to take a look at the implementation now, and I'm wondering if there isn't a more simple way to solve this problem by:

1. Create a "dummy" Unboxable struct, with the "real" unboxable nested within it.
2. Unbox the dummy struct with the given dictionary
3. Retrieve and return the nested dictionary.

Like this:

```swift
private struct UnboxContainerContext {
    let key: String
    let isKeyPath: Bool
}

private struct UnboxContainer<T: Unboxable>: UnboxableWithContext {
    let model: T
    
    init(unboxer: Unboxer, context: UnboxContainerContext) {
        self.model = unboxer.unbox(context.key, isKeyPath: context.isKeyPath)
    }
}

public func Unbox<T: Unboxable>(dictionary: UnboxableDictionary, at key: String, isKeyPath: Bool = true) throws -> T {
    let context = UnboxContainerContext(key: key, isKeyPath: isKeyPath)
    let container: UnboxContainer<T> = try Unbox(dictionary, context: context)
    return container.model
}
```

Let me know what you think :)

File: README.md (hunk 1 above)
Comment (clayellis) [inline README.md]: I'm sorry about that—I had opened the README in iA Writer (my markdown editor of choice) and the tabbing was different and it just got messed up. I didn't intend to change that.

File: README.md (hunk 2 above)
Comment (clayellis) [inline README.md]: Yeah, like I mentioned in my previous comment, tabbing in iA Writer was a bit different.

**Comment by clayellis** [comment]
@JohnSundell much better! I really like that solution. The usage looks like it would be the exact same from what I can tell:

```swift
let object: Object = try Unbox(json, at: "objectKey")
```

I'll remove my `unbox(...at:...)` method.

**Comment by clayellis** [comment]
@JohnSundell And would you like me to implement the feature you suggested?

**Comment by JohnSundell** [comment]
@clayellis You can handle the array case just as with the non-array case, make a container for it as well (like `UnboxArrayContainer`, for example). Feel free to implement this if you want, or let me know if you want me to implement it :) With this solution I don't think you need the `retrieveValues` methods etc.

File: README.md (hunk 1 above)
Comment (JohnSundell) [inline README.md]: Yeah, suspected that, no worries :) Used to happen to me all the time too (I love iA Writer, use it for everything, even my shopping lists 😜), then I discovered that you can use spaces for indentation in iA Writer as well 🎉

**Comment by clayellis** [comment]
@JohnSundell, great, I'll be ready to commit and push in about 5 minutes.

**Comment by JohnSundell** [comment]
@clayellis Awesome! I'll await your push sir! 👍 (But no stress, take your time 😀)

**Comment by clayellis** [comment]
@JohnSundell Pushed! I removed the old methods, restored `retrieveValues ` and added an additional check inside of that method to see if the index from the keyPath was out of range (necessary due to a failed test), removed my old methods, implemented yours, removed  an unneeded test, updated the README and (hopefully) fixed those indentations.

File: Tests/UnboxTests.swift [hunk 4]
```diff
@@ -1210,7 +1210,120 @@ class UnboxTests: XCTestCase {
...
+    func testUnboxingStartingAtMissingCustomKey() {
+        let dictionary: UnboxableDictionary = [
+            "A": [
+                "int": 14
+            ]
+        ]
+        
+        do {
+            let unboxed: UnboxTestSimpleMock = try Unbox(dictionary, at: "B")
+            XCTAssertEqual(unboxed.int, 14)
```
Comment (JohnSundell) [inline Tests/UnboxTests.swift]: You'd want to do a `XCTFail()` here, because the test has failed if the method doesn't throw.

Comment (JohnSundell) [inline Tests/UnboxTests.swift]: Same for the other similar pieces of code :)

**Comment by JohnSundell** [comment]
@clayellis Awesome!! 🎉 Small comment on the test code, after that's resolved I'll merge 🚀

File: Tests/UnboxTests.swift (hunk 4 above)
Comment (clayellis) [inline Tests/UnboxTests.swift]: D'oh! Right, hang tight...

**Comment by clayellis** [comment]
@JohnSundell 3.. 2.. 1.. 🚀

**Comment by JohnSundell** [comment]
Thanks so much for this @clayellis, really awesome! 🚀

**Comment by clayellis** [comment]
@JohnSundell for sure! Thanks for Unbox! I love using it and am looking forward to finding new ways to contribute!
